### ISSUE TITLE
Add `functional` prop

### ISSUE THREAD

**Issue by jmeas** [jmeas/react-composer#19]
`functional` would give developers more control over the `components`.

tl;dr, the result of each component is passed into the next. This would allow for async render prop components to work in series based off of the results of the previous component.

---

**Update** The name "serial" doesn't quite communicate how this works, because it doesn't enforce serial rendering, it just _enables_ it.

Here's an example of what the API might look like:

```jsx
<Composer functional components={[
    () => <RenderPropComponent {...configOne} />,
    (resultOne) => <RenderPropComponent {...configTwo} />,
    (resultOne, resultTwo) => <RenderPropComponent {...configThree} />
  ]}>
  {([resultOne, resultTwo, resultThree]) => (
    <MyComponent results={[resultOne, resultTwo, resultThree]} />
  )}
</Composer>
```

**Comment by erikthedeveloper** [comment]
Oh man 😅. I just now realized that this issue existed. I just created a similar issue over here https://github.com/jmeas/react-composer/issues/28 I would be happy to close it as a duplicate and and post the contents here as a comment if that help @jmeas.

**Comment by erikthedeveloper** [comment]
Using the results of an outer render prop component an inner render prop component is something I commonly need when manually composing render prop components so I'm excited to see it being discussed here 😃 

Some thoughts I have around the idea:

### Keeping signature consistent

Where the primary `Composer.props.children` signature is 

```js
([resultOne, resultTwo]) => {/* ... */}
```

I feel like it would be more consistent to follow that for the `Composer.props.components` entries as functions also. 

So rather than 

```js
[
  (resultOne, resultTwo) => {/* ... */},
  (resultOne) => {/* ... */},
  () => {/* ... */},
]
```

It would be

```js
[
  ([resultOne, resultTwo]) => {/* ... */},
  ([resultOne]) => {/* ... */},
  ([]) => {/* ... */},
]
```

### Support both `element` and `function` without requiring opt-in flag

Part of me likes the explicit `props.functional` from @jmeas example, but part of me wonders if both `element` and `function` could both be supported via a type check so this is a first class API rather than an opt-in alternative API. 

Proposed API example usage (_assuming a shift to first:Outer last:Inner as being discussed in #7_)

```jsx
<Composer
  components={[
    <Outer {...outerProps} />,
    <Middle {...middleProps} />,
    ([outerResult, middleResult]) => <InnerNeedsResults {...outerResult} {...middleResult} />,
  ]}>
  {([outerResult, middleResult, innerResult]) => (
    <MyComponent results={[outerResult, middleResult, innerResult]} />
  )}
</Composer>
```

As for how to achieve this, I roughly played with this idea after looking at this section of `chainComponents`

https://github.com/jmeas/react-composer/blob/02d11f844e40e09633fd2f069e1d96cd8339d410/src/index.js#L43-L52

and came up with this (totally untested ⚠️ 😄)

```js
let component = components[componentIndex];

// ... 

if (typeof component === 'function') {
  // This entry from props.components is a function!
  // Invoke it with the currently accumulated results to produce the element.
  component = component(results);
}

// component is an element as it always was. Business as usual...
return React.cloneElement(component, {/* ... */});
```

🤔  on a related note, I feel like naming (internally and externally) might warrant discussion where these are being called "components" throughout where they are really "elements" and component/element have a rather specific meaning in React https://reactjs.org/blog/2015/12/18/react-components-elements-and-instances.html

---

:memo: As I was thinking about this and writing this, I realized just how confusing it can be with either RTL or LTR ordering. I might suggest using language such as "outer", "middle" , "inner" in the README and examples. That just feels more explicit/clear to me and eliminates a bit of mental parsing (especially for those not super familiar with composition in general and especially within React).

**Comment by jmeas** [comment]
> I feel like it would be more consistent to follow that for the Composer.props.components entries as functions also.

:+1: That makes sense. I was kind of on the fence about this one for whatever reason, but I think consistency is a good idea here.

> Part of me likes the explicit props.functional from @jmeas example, but part of me wonders if both element and function could both be supported via a type check so this is a first class API rather than an opt-in alternative API.

This is a really great idea.

**Comment by erikthedeveloper** [comment]
@jmeas I'm unsure of what you mean by

> My only concern is that it seems to **lose support for function render prop components**.

My apologies if I'm totally missing something here, but are you imagining this would prevent usages such as ⬇️ ?

```jsx
function SomeFunctionRenderPropComponent(props) {
  return props.children(/* ... */);
}

 <Composer
  components={[
    // Equivalent to React.createElement(SomeFunctionRenderProp, someProps)
    <SomeFunctionRenderProp {...someProps} />,
  ]}>
  {(results) => {/* ... */}}
</Composer>
```

Whether your components are "function components" or "class components", by the time they make it into the `props.components` array they are no longer components, but rather _evaluated components_ which are then _elements_. So by adding this `typeof props.components[n] === 'function'` check we would essentially be asking "Is this already an element or is this a function [element factory]?".

From https://reactjs.org/blog/2015/12/18/react-components-elements-and-instances.html#components-can-be-classes-or-functions

> **Components Can Be Classes or Functions**
In the code above, Form, Message, and Button are React components. They can either be written as functions or as classes descending from React.Component. [...] However, whether functions or classes, fundamentally **they are all components** to React. They take the props as their input, and **return** the **elements** as their output.

This is related to what I was getting at with:

> I feel like naming in general might warrant discussion where these are being called "components" throughout where they are really "elements" both having a rather specific meaning in React

**Comment by jmeas** [comment]
> Whether your components are "function components" or "class components", by the time they make it into the props.components array they are no longer components, but rather evaluated components which are then elements.

Ah, right, right. Makes sense :+1:

> I feel like naming in general might warrant discussion where these are being called "components" throughout where they are really "elements" both having a rather specific meaning in React

I agree with you that this is worth discussing. Would you want to open a new issue for this?

---

Given that my only concern with supporting the API you proposed turned out to be a non-issue, I'm thinking we could move ahead and get a PR going for that update. If you have time to do that, I'd be happy to review it. I think the API you proposed + resolving #7 could be implemented at the same time. No worries if you're busy – I can get to it, but it may not be a little while.

Thanks for taking the time to share your thoughts here @erikthedeveloper !

**Comment by jmeas** [comment]
Resolved in #29 . Will be released as 3.1.0 shortly.
